cpp/src/arrow/array/builder_union.cc

```diff
@@ -19,42 +19,90 @@
 
 #include <utility>
 
+#include "arrow/util/checked_cast.h"
 #include "arrow/util/logging.h"
 
 namespace arrow {
 
-DenseUnionBuilder::DenseUnionBuilder(MemoryPool* pool,
-                                     const std::shared_ptr<DataType>& type)
-    : ArrayBuilder(type, pool), types_builder_(pool), offsets_builder_(pool) {}
+using internal::checked_cast;
 
-Status DenseUnionBuilder::FinishInternal(std::shared_ptr<ArrayData>* out) {
-  std::shared_ptr<Buffer> types;
+Status BasicUnionBuilder::FinishInternal(std::shared_ptr<ArrayData>* out) {
+  std::shared_ptr<Buffer> types, null_bitmap;
+  RETURN_NOT_OK(null_bitmap_builder_.Finish(&null_bitmap));
   RETURN_NOT_OK(types_builder_.Finish(&types));
-  std::shared_ptr<Buffer> offsets;
-  RETURN_NOT_OK(offsets_builder_.Finish(&offsets));
 
-  std::shared_ptr<Buffer> null_bitmap;
-  RETURN_NOT_OK(null_bitmap_builder_.Finish(&null_bitmap));
+  // If the type has not been specified in the constructor, gather type_codes
+  std::vector<uint8_t> type_codes;
+  if (type_ == nullptr) {
+    for (size_t i = 0; i < children_.size(); ++i) {
+      if (type_id_to_children_[i] != nullptr) {
+        type_codes.push_back(static_cast<uint8_t>(i));
+      }
+    }
+  } else {
+    type_codes = checked_cast<const UnionType&>(*type_).type_codes();
+  }
 
-  std::vector<std::shared_ptr<Field>> fields;
-  std::vector<std::shared_ptr<ArrayData>> child_data(children_.size());
-  std::vector<uint8_t> type_ids;
+  std::vector<std::shared_ptr<ArrayData>> child_data(type_codes.size());
```

Comment:
Shouldn't you pre-allocate to children_.size()?

Reply:
The two should have identical size at this point, but I can change it for clarity.